hal/src/thumbv6m/sercom/v2/uart.rs

```diff
@@ -0,0 +1,2024 @@
+//!
+//! ```
+//! use atsamd_hal::sercom::v2::uart::Uart;
+//! ```
+//!
+//! # Disabling and reconfiguring
+//!
+//! Some methods, such as [`disable`] and [`reconfigure`], need to operate on
+//! all parts of a UART at once. In practice, this means that these methods
+//! operate on the type that was returned by [`enable`]. This can be [`UartRx`],
+//! [`UartTx`], or a `([UartRx], [UartTx])` tuple, depending on how the
+//! peripheral was configured. Note that like [`enable`], [`disable`] and
+//! [`reconfigure`] need the [`Disable`] trait to be in scope.
+//!
+//! ```
+//! use atsamd_hal::sercom::v2::uart::{Enable, Disable};
+//! // Assume config is a valid UART Config struct
+//! let (rx, tx) = config.enable();
+//!
+//! // Send/receive data with tx/rx...
+//!
+//! (rx, tx).reconfigure(|c| c.baud(1.mhz()));
+//! ```
+//!
+//! # Using UART with DMA
+//!
+//! This HAL includes support for DMA-enabled UART transfers. [`UartRx`] and
+//! [`UartTx`] both implement the DMAC [`Buffer`](crate::dmac::transfer::Buffer)
+//! trait. The provided [`send_with_dma`](UartTx::send_with_dma) and
+//! [`receive_with_dma`](UartRx::receive_with_dma) build and begin a
+//! [`dmac::Transfer`](crate::dmac::transfer::Transfer), thus starting the UART
+//! in a non-blocking way. Optionally, interrupts can be enabled on the provided
+//! [`Channel`](crate::dmac::channel::Channel). Note that the `dma` feature must
+//! be enabled. Please refer to the [`dmac`](crate::dmac) module-level
+//! documentation for more information.
+//!
+//! ```
+//! // Assume channel0 and channel1 are configured `dmac::Channel`s,
+//! // rx is a UartRx, and tx is a UartTx.
+//!
+//! // Create data to send
+//! let tx_buffer: [u8; 50] = [0xff; 50];
+//! let rx_buffer: [u8; 100] = [0xab; 100];
+//!
+//! // Launch transmit transfer
+//! let tx_dma = tx.send_with_dma(&mut tx_buffer, channel0, ());
+//!
+//! // Launch receive transfer
+//! let rx_dma = rx.receive_with_dma(&mut rx_buffer, channel1, ());
+//!
+//! // Wait for transfers to complete and reclaim resources
+//! let (chan0, tx_buffer, tx) = tx_dma.wait();
+//! let (chan1, rx, rx_buffer) = rx_dma.wait();
+//! ```
+//!
+//! # Non-supported advanced features
+//!
+//! * Synchronous mode (USART) is not supported
+//!
+//! [`enable`]: Enable::enable
+//! [`disable`]: Disable::disable
+//! [`reconfigure`]: Disable::reconfigure
+//! [`Pin`]: crate::gpio::v2::pin::Pin
+//! [`PinId`]: crate::gpio::v2::pin::PinId
+//! [`OptionalPinId`]: crate::gpio::v2::pin::OptionalPinId
+
+use core::convert::{TryFrom, TryInto};
+use core::marker::PhantomData;
+
+use crate::target_device as pac;
+use crate::time::Hertz;
+use pac::{
+    sercom0::{usart::ctrla::MODE_A, RegisterBlock},
+    PM,
+};
+
+use crate::gpio::v2::{AnyPin, OptionalPin, SomePin};
+use crate::sercom::v2::*;
+use crate::typelevel::{Is, NoneT, Sealed};
+
+use bitflags::bitflags;
+use embedded_hal::blocking;
+use embedded_hal::serial::{Read, Write};
+use nb::Error::WouldBlock;
```

Comment:
I believe this needs to be `pub use nb::Error::WouldBlock;`, so that callers of `.read()` and `.write()` can match on the Result.

Reply:
I'd prefer if we reexported `nb` at the crate level instead (if we don't already do that? Maybe in the prelude?) since a lot of peripherals use e-hal blocking traits

Reply:
I may be missing something, but couldn't find a file in the HAL that made `nb` public.  So, I've made https://github.com/atsamd-rs/atsamd/pull/458